### PR TITLE
✨ feat: add Quantity API

### PR DESCRIPTION
**Pull request by nstarman**
This PR adds 2 runtime-checkable protocols that define the minimum Q API and the more complete API that includes the Array API. These protocols are made public in a new module named `api`.

private protocols and API are defined in the private modules `_src/array_api` and `_src/api`. The former is a temporary module that we can deprecate when the https://github.com/data-apis/array-api-typing package is developed and released.

Please squash-merge.

### CODE REVIEW

File: src/quantity/_src/array_api.py
Comment (nstarman) [inline src/quantity/_src/array_api.py:24]: I think mypyc has fixed the `Self` annotation problem. Should probably change this to `Self`, then work on getting mypyc set up.

File: src/quantity/_src/array_api.py
Comment (nstarman) [inline src/quantity/_src/array_api.py:8]: Then we could remove this.

**Comment by nstarman** [comment]
@mhvk are you good to ignore RUF022? The `__all__` are currently topically sorted.

**Comment by mhvk** [comment]
> @mhvk are you good to ignore RUF022? The `__all__` are currently topically sorted.

Sure, the sorting only starts to make sense when there are lots of entries. I don't think we're going to get there!

File: src/quantity/_src/array_api.py
Comment (mhvk) [reply src/quantity/_src/array_api.py:24]: Do we actually need `__pow__`? `__array_namespace__` should be enough, no? If we do think more is needed, I'd prefer to add something like `shape` and `dtype`.

File: src/quantity/__init__.py
Comment (mhvk) [inline src/quantity/__init__.py:11]: If we expose it, I think we should add tests...

**Review by mhvk — COMMENTED** [review]
This looks good, though with small queries inside. I do think if we expose anything, we should add tests for it. But I'm happy not to expose this yet, and instead just use it in the typing PR.

File: src/quantity/_src/array_api.py
Comment (nstarman) [reply src/quantity/_src/array_api.py:24]: Ideally `Array` would define the full gamut from https://data-apis.org/array-api/latest/API_specification/array_object.html. Otherwise static typers wouldn't know that you can e.g. add 2 quantities together.

Comment (nstarman) [reply src/quantity/_src/array_api.py:24]: I had `__pow__` because that's all that was needed at the time I started PR #8.

Comment (mhvk) [reply src/quantity/_src/array_api.py:24]: Not sure I fully understand, but the top is pretty clear that we hope the array API folds themselves will give us this eventually, so should be fine.

File: src/quantity/__init__.py
Comment (nstarman) [reply src/quantity/__init__.py:11]: added!

File: src/quantity/_src/array_api.py
Comment (nstarman) [reply src/quantity/_src/array_api.py:25]: > Should probably change this to Self, then work on getting mypyc set up.

I'd actually rather defer and see what happens with mypyc.

File: src/quantity/_src/api.py
Comment (mhvk) [inline src/quantity/_src/api.py:55]: Curious: why the second check? Can there be something in `__annotations__` that is not an attribute?

File: tests/test_api.py
Comment (mhvk) [inline tests/test_api.py]: Typo: isinstance

File: tests/test_api.py
Comment (mhvk) [inline tests/test_api.py:15]: How about we also check `u.Quantity`? It should work and we probably want it to continue to work.

Maybe also good to do `assert not issubclass(np.ndarray, ...`

File: tests/test_api.py
Comment (mhvk) [inline tests/test_api.py]: As above, also for `u.Quantity`?  And for different array types?

**Review by mhvk — APPROVED** [review]
Nice to have some tests. I  made some in-line comments about testing `u.Quantity` and the various array types too, but then thought it was probably easier to just push an extra commit...

So, that leaves only a question about the `__subclass_hook__` implementation. Will approve assuming that it is OK.

File: src/quantity/_src/api.py
Comment (nstarman) [reply src/quantity/_src/api.py:55]: classes don't need to have an `__annotations__` and particularly the value.
In PEP 544 GVR says "isinstance() can be used with both data and non-data protocols, while issubclass() can be used only with non-data protocols. This restriction exists because some data attributes can be set on an instance in constructor and this information is not always available on the class object". This is a way around that for common cases, like property descriptors or dataclasses (via the annotation).

File: tests/test_api.py
Comment (nstarman) [reply tests/test_api.py:14]: Done.

File: tests/test_api.py
Comment (nstarman) [reply tests/test_api.py]: corrected.

**Comment by nstarman** [comment]
@mhvk I refactored the tests to use pytest fixtures. They should be the same, just avoiding playing with `globals`.

File: src/quantity/_src/api.py
Comment (nstarman) [reply src/quantity/_src/api.py:55]: We could hold off on adding these methods. Then `api.Quantity` can't be used for `issubclass` checks, only `isinstance` checks...

File: tests/test_api.py
Comment (nstarman) [reply tests/test_api.py:34]: Alternatively we can just make a class-scoped fixture (https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#scope-sharing-fixtures-across-classes-modules-packages-or-session) by moving `def array_and_quantity(request):` into the class and adding the argument to the fixture. It neatly replaces `class_setup`.

**Comment by nstarman** [comment]
Plz squash-merge, editing the comment.

File: tests/test_api.py
Comment (nstarman) [reply tests/test_api.py]: try running pytest in verbose mode, it now repeats the whole class as a block for each array library.

**Review by mhvk — APPROVED** [review]
Not completely sold on the fixture for the general case still, but here it is very obvious what it does, so let's just get it in. Thanks, @nstarman!

**Pull request merged**
